common/predictive-text/build.sh

```diff
@@ -20,92 +20,147 @@ THIS_SCRIPT="$(greadlink -f "${BASH_SOURCE[0]}" 2>/dev/null || readlink -f "${BA
+  else
+    do_configure
+    CONFIGURED=configure:browser
+  fi
+  builder_finish_action success configure :browser
+fi
+
+if builder_start_action configure :headless; then
+  if [ -n "$CONFIGURED" ]; then
+    echo "Configuration already completed in ${BUILDER_TERM_START}${CONFIGURED}${BUILDER_TERM_END}; skipping."
+  else
+    do_configure
+    CONFIGURED=configure:headless
+  fi
+
+  builder_finish_action success configure :headless
+fi
+
+### CLEAN ACTIONS
+
 # A nice, extensible method for -clean operations.  Add to this as necessary.
-clean ( ) {
+do_clean() {
   if [ -d $LMLAYER_OUTPUT ]; then
-    rm -rf "$LMLAYER_OUTPUT" || fail "Failed to erase the prior build."
+    rm -rf "$LMLAYER_OUTPUT"
   fi
 }
 
-display_usage ( ) {
-  echo "Usage: $0 [-clean] [-skip-package-install | -S] [-test | -tdd]"
-  echo "       $0 -help"
-  echo
-  echo "  -clean                 to erase pre-existing build products before a re-build"
-  echo "  -help                  displays this screen and exits"
-  echo "  -skip-package-install  (or -S) skips dependency updates"
-  echo "  -tdd                   skips dependency updates, builds, then runs unit tests only"
-  echo "  -test                  runs unit and integration tests after building"
-}
-
-################################ Main script ################################
+CLEANED=
+if builder_start_action clean :libraries; then
+  do_clean
+  CLEANED=clean:libraries
 
-run_tests=0
-fetch_deps=true
-unit_tests_only=0
-
-# Process command-line arguments
-while [[ $# -gt 0 ]] ; do
-  key="$1"
-  case $key in
-    -clean)
-      clean
-      ;;
-    -help|-h)
-      display_usage
-      exit
-      ;;
-    -skip-package-install|-S)
-      fetch_deps=false
-      ;;
-    -test)
-      run_tests=1
-      ;;
-    -tdd)
-      run_tests=1
-      fetch_deps=false
-      unit_tests_only=1
-      ;;
-    *)
-      echo "$0: invalid option: $key"
-      display_usage
-      exit $EX_USAGE
-  esac
-  shift # past the processed argument
-done
-
-# Check if Node.JS/npm is installed.
-verify_npm_setup $fetch_deps
-
-if $fetch_deps; then
-  # We need to build keyman-version and lm-worker with a script for now
-  "$KEYMAN_ROOT/common/web/keyman-version/build.sh" || fail "Could not build keyman-version"
-  "$KEYMAN_ROOT/common/web/lm-worker/build.sh" || fail "Could not build lm-worker"
+  builder_finish_action success clean :libraries
 fi
 
-build-browser || fail "Browser-oriented compilation failed."
-build-headless || fail "Headless compilation failed."
-echo "Typescript compilation successful."
+if builder_start_action clean :browser; then
+  if [ -n "$CLEANED" ]; then
+    echo "${BUILDER_TERM_START}clean${BUILDER_TERM_END} already completed as ${BUILDER_TERM_START}${CLEANED}${BUILDER_TERM_END}; skipping."
+  else
+    do_clean
+    CLEANED=clean:browser
+  fi
+  builder_finish_action success clean :browser
+fi
```

Comment:
It's slightly problematic that `clean` of one target also cleans the other target. Is there any way we can keep them isolated?

Reply:
Yes, separate folders is the correct answer, but agree it's a bit of a yak shave that doesn't need to happen here. Let's look at this as a future maintenance task.